infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/receive/PersistedWriterMetadata.java

```diff
@@ -0,0 +1,268 @@
+            log.warn("The metadata is older than the presisted one. Set snapshotStart siteConfigID " + siteConfigID + " ts " + ts +
+                    " persistSiteConfigID " + persistSiteConfigID + " persistSnapStart " + persistSnapStart);
+            return;
+        }
+
+        TxBuilder txBuilder = corfuStore.tx(namespace);
+
+        //Update the siteConfigID to fence all other transactions that update the metadata at the same time
+        appendUpdate(txBuilder, PersistedWriterMetadataType.SiteConfigID, siteConfigID);
+
+        //Setup the LastSnapStart
+        appendUpdate(txBuilder, PersistedWriterMetadataType.LastSnapStart, ts);
+
+        //Setup LastSnapSeqNum
+        appendUpdate(txBuilder, PersistedWriterMetadataType.LastSnapSeqNum, Address.NON_ADDRESS);
+        txBuilder.commit(timestamp);
+
+        log.debug("Commit. Set snapshotStart siteConfigID " + siteConfigID + " ts " + ts +
+                " persistSiteConfigID " + persistSiteConfigID + " persistSnapStart " + persistSnapStart);
+
+        return;
+    }
+
+
+    /**
+     * This call should be done in a transaction after a transfer done and before apply the snapshot.
+     * @param ts
+     */
+    public void setLastSnapTransferDoneTimestamp(long siteConfigID, long ts) {
+        CorfuStoreMetadata.Timestamp timestamp = corfuStore.getTimestamp();
+        long persisteSiteConfigID = query(timestamp, PersistedWriterMetadataType.SiteConfigID);
+        long persistSnapStart = query(timestamp, PersistedWriterMetadataType.LastSnapStart);
+
+        log.debug("setLastSnapTransferDone snapshotStart siteConfigID " + siteConfigID + " ts " + ts +
+                " persisteSiteConfigID " + persisteSiteConfigID + " persistSnapStart " + persistSnapStart);
+
+        // It means the site config has changed, ingore the update operation.
+        if (siteConfigID != persisteSiteConfigID || ts <= persisteSiteConfigID) {
+            log.warn("The metadata is older than the presisted one. Set snapshotStart siteConfigID " + siteConfigID + " ts " + ts +
+                    " persisteSiteConfigID " + persisteSiteConfigID + " persistSnapStart " + persistSnapStart);
+            return;
+        }
+
+        TxBuilder txBuilder = corfuStore.tx(namespace);
+
+        //Update the siteConfigID to fence all other transactions that update the metadata at the same time
+        appendUpdate(txBuilder, PersistedWriterMetadataType.SiteConfigID, siteConfigID);
+
+        //Setup the LastSnapStart
+        appendUpdate(txBuilder, PersistedWriterMetadataType.LastSnapTransferDone, ts);
+
+        txBuilder.commit(timestamp);
+
+        log.debug("Commit. Set snapshotStart siteConfigID " + siteConfigID + " ts " + ts +
+                " persisteSiteConfigID " + persisteSiteConfigID + " persistSnapStart " + persistSnapStart);
+        return;
+    }
+
+    public void setSrcBaseSnapshotDone(LogReplicationEntry entry) {
```

Comment:
comments

Reply:
The transfer done and apply done steps need to be documented at a high level class. A high level flow diagram in ascii would be very useful.